distributed/protocol/serialize.py

```diff
@@ -71,29 +74,111 @@ def pickle_loads(header, frames):
     return pickle.loads(x, buffers=buffers)
 
 
+def import_allowed_module(name):
+    if name in _cached_allowed_modules:
+        return _cached_allowed_modules[name]
+
+    # Check for non-ASCII characters
+    name = name.encode("ascii").decode()
+    # We only compare the root module
+    root = name.split(".", 1)[0]
+
+    # Note, if an empty string creeps into allowed-imports it is disallowed explicitly
+    if root and root in dask.config.get("distributed.scheduler.allowed-imports"):
+        _cached_allowed_modules[name] = importlib.import_module(name)
+        return _cached_allowed_modules[name]
+    else:
+        raise RuntimeError(
+            f"Importing {repr(name)} is not allowed, please add it to the list of "
+            "allowed modules the scheduler can import via the "
+            "distributed.scheduler.allowed-imports configuration setting."
+        )
+
+
 def msgpack_decode_default(obj):
     """
-    Custom packer/unpacker for msgpack to support Enums
+    Custom packer/unpacker for msgpack
     """
     if "__Enum__" in obj:
-        mod = importlib.import_module(obj["__module__"])
-        enum_type = getattr(mod, obj["__name__"])
-        obj = getattr(enum_type, obj["name"])
+        mod = import_allowed_module(obj["__module__"])
+        typ = getattr(mod, obj["__name__"])
+        return getattr(typ, obj["name"])
+
+    if "__Set__" in obj:
+        return set(obj["as-list"])
+
+    if "__SubgraphCallable__" in obj:
+        mod = import_allowed_module(obj["__module__"])
+        typ = getattr(mod, obj["__name__"])
+        return typ(*obj["args"])
+
+    if "__Layer__" in obj:
+        obj_name = obj["__name__"]
+        if obj_name == "BasicLayer":
+            # The default implemention of Layer returns a BasicLayer, which might
+            # not be defined in `mod` therefore we import it explicitly here
+            typ = BasicLayer
+        else:
+            mod = import_allowed_module(obj["__module__"])
+            typ = getattr(mod, obj["__name__"])
+        return typ(*obj["args"])
+
+    if "__HighLevelGraph__" in obj:
+        return HighLevelGraph(
+            obj["layers"],
+            obj["dependencies"],
+        )
+
+    if "__Serialized__" in obj:
+        return Serialized(*obj["data"])
```

Comment:
Nice! Just wanted to confirm with you that we're using `Serialized` here to avoid deserializing data which could cause us to execute arbitrary code with `pickle`

Reply:
Yes, I have added:
```python 
    if "__Serialized__" in obj:
        # Notice, the data here is marked a Serialized rather than deserialized. This
        # is because deserialization requires Pickle which the Scheduler cannot run
        # because of security reasons.
        # By marking it Serialized, the data is passed through to the workers that
        # eventually will deserialize it.
        return Serialized(*obj["data"])
```

Reply:
Great, thanks for taking the time to write up a comment : )